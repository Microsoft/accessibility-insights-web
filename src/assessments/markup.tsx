// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import * as React from 'react';

export function Tag(props: { tagName: string; isBold?: boolean }): JSX.Element {
    const isBold = props.hasOwnProperty('isBold') ? props.isBold : true;
    return isBold ? <CodeTerm>&lt;{props.tagName}&gt;</CodeTerm> : <Code>&lt;{props.tagName}&gt;</Code>;
}

export function Emphasis(props: { children: React.ReactNode }): JSX.Element {
    return <i>{props.children}</i>;
}

export function Term(props: { children: React.ReactNode }): JSX.Element {
    return <b>{props.children}</b>;
}

export function Code(props: { children: React.ReactNode }): JSX.Element {
    return <span className="insights-code">{props.children}</span>;
}

export const GreaterThanOrEqualTo = () => <span>&#8805;</span>;

export const NonBreakingSpace = () => <span>&nbsp;</span>;

export function CodeTerm(props: { children: React.ReactNode }): JSX.Element {
    return (
        <Term>
            <Code>{props.children}</Code>
        </Term>
    );
}
