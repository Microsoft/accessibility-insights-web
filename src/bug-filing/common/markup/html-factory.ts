// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { escape } from 'lodash';
import { MarkupFactory } from './markup-factory';
import { truncateSnippet as truncate } from './truncate-snippet';

export const createFactory = (truncateSnippet: (text: string) => string): MarkupFactory => {
    const bold = (text: string): string => {
        return `<b>${text}</b>`;
    };

    const snippet = (text: string): string => {
        const truncated = truncateSnippet(text);

        return `<code>${escape(truncated)}</code>`;
    };

    const link = (href: string, text?: string): string => {
        return `<a href="${href}">${text || href}</a>`;
    };

    const sectionSeparator = () => {
        return '<br><br>';
    };

    const newLine = () => {
        return '<br>';
    };

    const howToFixSection = (failureSummary: string): string => {
        return escape(failureSummary)
            .replace(/\n  /g, '<br>- ')
            .replace(/\n /g, '<br> ')
            .replace(/\n/g, '<br>');
    };

    return {
        bold,
        snippet,
        link,
        sectionSeparator,
        newLine,
        howToFixSection,
    };
};

export const HTMLFactory = createFactory(truncate);
