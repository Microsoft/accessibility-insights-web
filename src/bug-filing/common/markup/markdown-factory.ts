// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { MarkupFactory } from './markup-factory';
import { truncateSnippet as truncate } from './truncate-snippet';

export const createFactory = (truncateSnippet: (text: string) => string): MarkupFactory => {
    const bold = (text: string): string => {
        return `**${text}**`;
    };

    const snippet = (text: string): string => {
        const truncated = truncateSnippet(text);

        return `\`${truncated}\``;
    };

    const link = (href: string, text?: string): string => {
        return `[${text || href}](${href})`;
    };

    const sectionSeparator = (): string => {
        return '';
    };

    const newLine = (): string => {
        return '';
    };

    const howToFixSection = (failureSummary: string): string => {
        return failureSummary
            .split('\n')
            .map(line => `    ${line}`)
            .join('\n');
    };

    return {
        bold,
        snippet,
        link,
        sectionSeparator,
        newLine,
        howToFixSection,
    };
};

export const MarkdownFactory = createFactory(truncate);
