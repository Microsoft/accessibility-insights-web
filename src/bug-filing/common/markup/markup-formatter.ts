// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

export type MarkupFormatter = {
    bold(text: string): string;
    snippet(text: string): string;
    link(href: string, text?: string): string;
    sectionSeparator(): string;
    newLine(): string;
    howToFixSection(failureSummary: string): string;
};
