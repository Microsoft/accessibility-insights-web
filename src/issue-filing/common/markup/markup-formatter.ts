// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { HowToFixInstructions } from '../../../injected/adapters/scan-results-to-unified-results';

export type MarkupFormatter = {
    snippet(text: string): string;
    link(href: string, text?: string): string;
    sectionHeader(text: string): string;
    howToFixSection(failureSummary: string): string;
    howToFix(instructions: HowToFixInstructions): string;
    sectionHeaderSeparator(): string;
    footerSeparator(): string;
    sectionSeparator(): string;
};
