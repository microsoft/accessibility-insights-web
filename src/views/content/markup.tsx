// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { Code, Emphasis, Tag, Term } from 'assessments/markup';
import { NewTabLink } from 'common/components/new-tab-link';
import { CheckIcon } from 'common/icons/check-icon';
import { CrossIcon } from 'common/icons/cross-icon';
import { ContentActionMessageCreator } from 'common/message-creators/content-action-message-creator';
import { TextContent } from 'content/strings/text-content';
import * as React from 'react';
import { Helmet } from 'react-helmet';
import { CodeExample, CodeExampleProps } from './markup/code-example';

type PassFailProps = {
    passText: JSX.Element;
    passExample?: React.ReactNode;
    failText: JSX.Element;
    failExample?: React.ReactNode;
};

export type Markup = {
    Tag: React.FC;
    Code: React.FC;
    Term: React.FC;
    Emphasis: React.FC;
    Do: React.FC;
    Dont: React.FC;
    Pass: React.FC;
    Fail: React.FC;
    PassFail: React.FC<PassFailProps>;
    Columns: React.FC;
    Column: React.FC;
    Inline: React.FC;
    HyperLink: React.FC<{ href: string }>;
    Title: React.FC<{ children: string }>;
    Highlight: React.FC;
    CodeExample: React.FC<CodeExampleProps>;
    Links: React.FC;
    Table: React.FC;
    LandmarkLegend: React.FC<{ role: string }>;
    ProblemList: React.FC;
    Include: React.FC<{ content: MarkupBasedComponent }>;
};

export type MarkupDeps = {
    textContent: Pick<TextContent, 'applicationTitle'>;
    contentActionMessageCreator: Pick<ContentActionMessageCreator, 'openContentHyperLink'>;
};

export interface MarkupOptions {
    setPageTitle?: boolean;
}

export type MarkupBasedComponentProps = {
    deps: MarkupDeps;
    options?: MarkupOptions;
};

export type MarkupBasedComponent = React.FC<MarkupBasedComponentProps>;

export const createMarkup = (deps: MarkupDeps, options?: MarkupOptions) => {
    function Include(props: { content: MarkupBasedComponent }): JSX.Element {
        const Content = props.content;
        return <Content deps={deps} options={options} />;
    }

    function Title(props: { children: string }): JSX.Element {
        const { applicationTitle } = deps.textContent;
        const helmet = (
            <Helmet>
                <title>
                    {props.children} - {applicationTitle}
                </title>
            </Helmet>
        );

        return (
            <>
                {options && options.setPageTitle && helmet}
                <h1>{props.children}</h1>
            </>
        );
    }

    function HyperLink(props: { href: string; children: React.ReactNode }): JSX.Element {
        const { href } = props;
        const { openContentHyperLink } = deps.contentActionMessageCreator;

        return (
            <NewTabLink href={href} onClick={e => openContentHyperLink(e, href)}>
                {props.children}
            </NewTabLink>
        );
    }

    function Links(props: { children: React.ReactNode }): JSX.Element {
        return (
            <>
                <div className="content-hyperlinks">
                    {React.Children.map(props.children, el => el)}
                </div>
            </>
        );
    }

    function Inline(props: { children: React.ReactNode }): JSX.Element {
        return <div className="content-inline">{props.children}</div>;
    }

    function Do(props: { children: React.ReactNode }): JSX.Element {
        return (
            <Column>
                <div className="do-header">
                    <h2>Do</h2>
                    <CheckIcon />
                </div>
                <div className="do-section">{props.children}</div>
            </Column>
        );
    }

    function Dont(props: { children: React.ReactNode }): JSX.Element {
        return (
            <Column>
                <div className="dont-header">
                    <h2>Don't</h2>
                    <CrossIcon />
                </div>
                <div className="dont-section">{props.children}</div>
            </Column>
        );
    }

    function Pass(props: { children: React.ReactNode }): JSX.Element {
        return (
            <Column>
                <div className="pass-header">
                    <CheckIcon /> <h3>Pass</h3>
                </div>
                <div className="pass-section">{props.children}</div>
            </Column>
        );
    }

    function Fail(props: { children: React.ReactNode }): JSX.Element {
        return (
            <Column>
                <div className="fail-header">
                    <CrossIcon /> <h3>Fail</h3>
                </div>
                <div className="fail-section">{props.children}</div>
            </Column>
        );
    }

    function LandmarkLegend(props: { role: string; children: React.ReactNode }): JSX.Element {
        return <span className={`landmarks-legend ${props.role}-landmark`}>{props.children}</span>;
    }

    function Highlight(props: { children: React.ReactNode }): JSX.Element {
        return <span className="highlight">{props.children}</span>;
    }

    function Table(props: { children: React.ReactNode }): JSX.Element {
        return <ul className="table">{props.children}</ul>;
    }

    function ProblemList(props: { children: React.ReactNode }): JSX.Element {
        return <ul className="accessibility-problems-list">{props.children}</ul>;
    }

    function Columns(props: { children: React.ReactNode }): JSX.Element {
        return <div className="columns">{props.children}</div>;
    }

    function Column(props: { children: React.ReactNode }): JSX.Element {
        return <div className="column">{props.children}</div>;
    }

    function PassFail(props: PassFailProps): React.ReactNode {
        const { passText, passExample, failText, failExample } = props;

        function formatExample(example: string | React.ReactNode): React.ReactNode {
            if (typeof example === 'string') {
                return <CodeExample>{example}</CodeExample>;
            }
            return example;
        }

        return (
            <div className="pass-fail-grid">
                <div className="fail-section">
                    <div className="fail-header">
                        <CrossIcon /> <h3>Fail</h3>
                    </div>
                    {failText}
                </div>
                {failExample && <div className="fail-example">{formatExample(failExample)}</div>}
                <div className="pass-section">
                    <div className="pass-header">
                        <CheckIcon /> <h3>Pass</h3>
                    </div>
                    {passText}
                </div>
                {passExample && <div className="pass-example">{formatExample(passExample)}</div>}
            </div>
        );
    }

    return {
        Tag,
        Code,
        Term,
        Emphasis,
        Do,
        Dont,
        Pass,
        Fail,
        PassFail,
        Columns,
        Column,
        Inline,
        HyperLink,
        Title,
        CodeExample,
        Highlight,
        Links,
        LandmarkLegend,
        Table,
        ProblemList,
        options,
        Include,
    } as Markup;
};
