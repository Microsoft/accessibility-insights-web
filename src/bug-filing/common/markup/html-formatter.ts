// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { escape } from 'lodash';
import { MarkupFormatter } from './markup-formatter';
import { truncateSnippet as truncate } from './truncate-snippet';

export const createFormatter = (truncateSnippet: (text: string) => string): MarkupFormatter => {
    const bold = (text: string): string => {
        return `<b>${text}</b>`;
    };

    const snippet = (text: string): string => {
        const truncated = truncateSnippet(text);

        return `<code>${escape(truncated)}</code>`;
    };

    const link = (href: string, text?: string): string => {
        return `<a href="${href}">${text || href}</a>`;
    };

    const sectionSeparator = () => {
        return '<br><br>';
    };

    const newLine = () => {
        return '<br>';
    };

    const howToFixSection = (failureSummary: string): string => {
        return escape(failureSummary)
            .replace(/\n  /g, '<br>- ')
            .replace(/\n /g, '<br> ')
            .replace(/\n/g, '<br>');
    };

    return {
        bold,
        snippet,
        link,
        sectionSeparator,
        newLine,
        howToFixSection,
    };
};

export const HTMLFormatter = createFormatter(truncate);
