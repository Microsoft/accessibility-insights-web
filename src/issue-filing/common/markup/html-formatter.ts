// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.import { escape, isEmpty } from 'lodash';
import { isEmpty } from 'lodash';

import { HowToFixInstructions } from '../../../injected/adapters/scan-results-to-unified-results';
import { buildHowToFixInstructions, InstructionListBuilder } from './how-to-fix-instruction-builder';
import { MarkupFormatter } from './markup-formatter';
import { truncateSnippet as truncate } from './truncate-snippet';

export const createFormatter = (
    truncateSnippet: (text: string) => string,
    howToFixInstructionBuilder: (instructions: HowToFixInstructions, listBuilder: InstructionListBuilder) => string,
): MarkupFormatter => {
    const snippet = (text: string): string => {
        const truncated = truncateSnippet(text);

        return `<code>${escape(truncated)}</code>`;
    };

    const link = (href: string, text?: string): string => {
        return `<a href="${href}">${text || href}</a>`;
    };

    const sectionHeader = (text: string) => {
        return `<h4>${text}</h4>`;
    };

    const howToFixSection = (failureSummary: string): string => {
        return escape(failureSummary)
            .replace(/\n  /g, '<br>- ')
            .replace(/\n /g, '<br> ')
            .replace(/\n/g, '<br>');
    };

    const howToFix = (instructions: HowToFixInstructions) => {
        return howToFixInstructionBuilder(
            instructions,
            (title, checks) => `<div class="insights-fix-instruction-title">${title}</div>
                                    <ul class="insights-fix-instruction-list">
                                        ${checks.map(check => `<li>${check}</li>`).join()}
                                    </ul>
                                `,
        );
    };

    const sectionHeaderSeparator = () => null;

    const footerSeparator = () => null;

    const sectionSeparator = () => '\n';

    return {
        snippet,
        link,
        sectionHeader,
        howToFixSection,
        sectionHeaderSeparator,
        footerSeparator,
        sectionSeparator,
    };
};

export const HTMLFormatter = createFormatter(truncate, buildHowToFixInstructions);
