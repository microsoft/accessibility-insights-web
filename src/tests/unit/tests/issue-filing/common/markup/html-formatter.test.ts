// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { IMock, It, Mock, MockBehavior } from 'typemoq';

import { createFormatter } from '../../../../../../issue-filing/common/markup/html-formatter';
import { MarkupFormatter } from '../../../../../../issue-filing/common/markup/markup-formatter';

describe('HTMLFormatter', () => {
    let truncateMock: IMock<(text: string) => string>;
    let testSubject: MarkupFormatter;

    beforeEach(() => {
        truncateMock = Mock.ofInstance((text: string) => text, MockBehavior.Strict);
        testSubject = createFormatter(truncateMock.object);
    });

    it('returns section header', () => {
        expect(testSubject.sectionHeader('test-header')).toEqual('<h4>test-header</h4>');
    });

    it('returns how to fix section', () => {
        const failureSummary = 'fix\n  1\n 2\n3\n4';

        const result = testSubject.howToFixSection(failureSummary);

        expect(result).toEqual('fix<br>- 1<br> 2<br>3<br>4');
    });

    it('returns section header separator', () => {
        expect(testSubject.sectionHeaderSeparator()).toBe(null);
    });

    it('returns footer separator', () => {
        expect(testSubject.footerSeparator()).toBe(null);
    });

    it('return new line', () => {
        expect(testSubject.sectionSeparator()).toBe('\n');
    });

    describe('creates snippet', () => {
        beforeEach(() => {
            truncateMock.setup(truncate => truncate(It.isAnyString())).returns(text => text);
        });

        it('no need to escape', () => {
            const result = testSubject.snippet('this is code');

            expect(result).toEqual('<code>this is code</code>');
        });

        it('escapes properly', () => {
            const result = testSubject.snippet('<div>text</div>');

            const expected = '<code>&lt;div&gt;text&lt;/div&gt;</code>';
            expect(result).toEqual(expected);
        });
    });

    describe('relatedPaths', () => {
        it('returns as an HTML list', () => {
            expect(testSubject.relatedPaths(['#path-a', '#path-b'])).toEqual(
                '<ul><li>#path-a</li><li>#path-b</li></ul>',
            );
        });

        it('escapes properly', () => {
            expect(testSubject.relatedPaths(['[attr="<script>"]'])).toEqual(
                '<ul><li>[attr=&quot;&lt;script&gt;&quot;]</li></ul>',
            );
        });
    });

    describe('creates link', () => {
        it('handles href only', () => {
            const result = testSubject.link('test-href');

            expect(result).toEqual('<a href="test-href">test-href</a>');
        });

        it('handles href and text', () => {
            const result = testSubject.link('test-href', 'test-text');

            expect(result).toEqual('<a href="test-href">test-text</a>');
        });
    });
});
