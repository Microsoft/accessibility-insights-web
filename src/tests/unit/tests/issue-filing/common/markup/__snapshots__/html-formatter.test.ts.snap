// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTMLFormatter returns how to fix section from HowToFixInstructions object 1`] = `
"<div class=\\"insights-fix-instruction-title\\">Fix ALL of the following: </div>
                                    <ul class=\\"insights-fix-instruction-list\\">
                                        <li>3</li>,<li>2</li>
                                    </ul>
                                <div class=\\"insights-fix-instruction-title\\">Fix the following:</div>
                                    <ul class=\\"insights-fix-instruction-list\\">
                                        <li>1</li>
                                    </ul>
                                "
`;
